docs/source/data.rst

```diff
@@ -2,11 +2,342 @@ torch.utils.data
+
+`Data Loading Strategies`_ section talks about how to use a :class:`~torch.utils.data.Sampler`
+with a :class:`~torch.utils.data.DataLoader`.
+
+Data Loading Strategies
+-----------------------
+
+:class:`~torch.utils.data.DataLoader` constructor receives a
+:attr:`dataset` object as its first argument. Based on the other provided
+arguments, a :class:`~torch.utils.data.DataLoader` operates in one of three
+following strategies:
+
+* `Batched loading from a map-style dataset (default)`_
+
+* `Loading individual members of a map-style dataset`_
+
+* `Loading from an iterable-style dataset`_
+
+Batched loading from a map-style dataset (default)
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+This is the most common case, and corresponds to fetching a minibatch of
+data and collating them into batched Tensors, i.e., Tensors with one dimension
+being the batch dimension (usually the first). Two combinations of arguments
+that starts this mode are:
+
+* (default) Using arguments :attr:`batch_size`, :attr:`shuffle`,
+  :attr:`sampler`, and :attr:`drop_last` to specify the batch indices sampling
+  behavior.
+
+  With the default arguments, a :class:`~torch.utils.data.DataLoader`
+  loads data as batches of size ``1`` with indices sampled without replacement.
+
+* Setting argument :attr:`batch_sampler` to a custom sampler returning a list
+  of indices at each time, representing the indices for a batch.
+
+After fetching a list of samples using the indices from sampler, the function
+passed as the :attr:`collate_fn` argument is used to collate the list of samples
+into batched Tensors.
+
+The behavior of this mode is roughly equivalent with::
+
+    for indices in batch_sampler:
+        yield collate_fn([dataset[i] for i in indices])
+
+
+Working with :attr:`collate_fn`
+"""""""""""""""""""""""""""""""
+
+For instance, if each data sample consists of a 3-channeled image and an
+integral class label, i.e., each element of the dataset returns a tuple
+``(image, class_index)``, the default :attr:`collate_fn` collates a list of such
+tuples into a single tuple of a batched image tensor and a batched class label
+Tensor. In particular, the default :attr:`collate_fn` has the following
+properties:
+
+* It always prepends a new dimension as the batch dimension.
+
+* It automatically converts NumPy arrays and Python numerical values into
+  PyTorch Tensors.
+
+* It preserves the data structure, e.g., if each sample is a dictionary, it
+  outputs a dictionary with the same set of keys but batched Tensors as values
+  (or lists if the values can not be converted into Tensors). Same
+  for ``list`` s, ``tuple`` s, ``namedtuple`` s, etc.
+
+Users may use customized :attr:`collate_fn` to achieve custom batching, e.g.,
+along a dimension other than the first.
+
+Loading individual members of a map-style dataset
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+Sometimes it is cheaper to directly load batched data (e.g., bulk reads from a
```

Comment:
This is a few feature of this PR, right? I'm not 100% convinced that we want to implement it this way. What you do here is that you really have a `BatchDataset`, which takes an index of a batch to return. This has a few problems like the fact that the division into batches does not depend on the sampler, and you need to reset that at every epoch too. Hence, I think that in this mode we should really call `__getitem__` with a list or a tuple of indices to return, allowing the dataset to perform a batched query.

Reply:
Also, then we can make the dataset return a list/tuple of answers to that query, and we don't need to add `convert_fn`, because it still corresponds to `collate_fn`

Reply:
Yes. You are right about what this mode does. I should word this part of the documentation better because I think the mode supports very similar to what you describes. Let me describe the use case I had in mind when implementing it, which I believe will make things clearer:

Users pass in a `sampler` that each time yields batched indices as a list (or as any type really, e.g., partition idx of a dataset in a database), and a `dataset` object whose `__getitem__` can handle such batch indices type, and will properly do bulk loading, and yield batched tensors. They also have the choice to either do batching themselves in `__getitem__` if their bulk loading returns data in nice contiguous chunks, or supply `default_collate` as the `convert_fn`.

Let me know if this makes sense to you. I'd be happy to hear your suggestions and opinions on it!

Finally, I want to note that `convert_fn` is still needed even if we kill / completely change this mode's behavior, as it is needed for the `IterableDataset` case.